File: #2471.minimum-number-of-operations-to-sort-a-binary-tree-by-level.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minimumOperations(TreeNode* root) {
        
        if (!root) {
            return 0;
        }

        int swaps = 0;
        queue<TreeNode *>q;
        q.push(root);
        while (!q.empty()) {
            int size = q.size();
            vector<int> v;
            while (size--) {
                auto p = q.front();
                q.pop();
                v.push_back(p->val);
                if (p->left) {
                    q.push(p->left);
                }

                if (p->right) {
                    q.push(p->right);
                }
            }

            vector<int> temp (v);
            unordered_map<int,int> mp;

            std::sort(temp.begin(), temp.end());
            
            size = temp.size();
            for (int i = 0; i < size; i++) {
                mp[temp[i]] = i;
            }

            for (int i = 0; i < size; i++) {
                if (mp[v[i]] != i) {
                    std::swap(v[i], v[mp[v[i]]]);
                    swaps++;
                    i--;
                }
            }
        }
        return swaps;
    }
};

File: #2471.minimum-number-of-operations-to-sort-a-binary-tree-by-level.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minimumOperations = function(root) {
    let queue = [root];
    let res = 0;
    while (queue.length) {
        const newQueue = [];
        const arr = [];
        let pos = [];
        const posNum = {};
        const numPos = {};
        for (let i = 0; i < queue.length; i++) {
            q = queue[i];
            arr.push([q.val, i]);
            pos.push(q.val);
            posNum[i] = q.val;
            numPos[q.val] = i;
            if (q.left) {
                newQueue.push(q.left);
            }
            if (q.right) {
                newQueue.push(q.right);
            }
        }

        pos.sort((a, b) => a - b);

        for (let i = 0; i < pos.length; i++) {
            if (posNum[i] === pos[i]) {
                continue;
            }
            res++;
            const numOnCurrPos = posNum[i];
            const moveNumOnPos = numPos[pos[i]];
            numPos[pos[i]] = i;
            posNum[i] = pos[i];
            numPos[numOnCurrPos] = moveNumOnPos;
            posNum[moveNumOnPos] = numOnCurrPos;
        }
        queue = newQueue;
    }
    return res;
};
